python/python/lance/arrow.py

```diff
@@ -11,14 +11,428 @@
+    def __init__(self, arrow_type, shape):
+        self.shape = shape
+        self.arrow_type = arrow_type
+        length = 1
+        for dim in shape:
+            length *= dim
+
+        pa.ExtensionType.__init__(
+            self,
+            pa.list_(arrow_type, length),
+            "lance.arrow.fixed_shape_image_tensor",
+        )
+
+    def __arrow_ext_serialize__(self):
+        import json
+
+        serialized = json.dumps(
+            {"shape": self.shape, "parent_type": "FixedShapeImageTensorType"}
+        ).encode()
+        return serialized
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        import json
+
+        deserialized = json.loads(serialized.decode())
+        return FixedShapeImageTensorType(storage_type.value_type, deserialized["shape"])
+
+    def __arrow_ext_class__(self):
+        return FixedShapeImageTensorArray
+
+    def __arrow_ext_scalar_class__(self):
+        return FixedShapeImageTensorScalar
+
+
+pa.register_extension_type(ImageURIType())
+pa.register_extension_type(EncodedImageType())
+pa.register_extension_type(FixedShapeImageTensorType(pa.uint8(), (0,)))
+
+
+class ImageArray(pa.ExtensionArray):
+    def __repr__(self):
+        return "<lance.arrow.%s object at 0x%016x>\n%s" % (
+            type(self).__name__,
+            id(self),
+            repr(self.to_pylist()),
+        )
+
+    @classmethod
+    def from_array(cls, images):
+        """
+        Create an one of subclasses of ImageArray from input data.
+
+        Parameters
+        ----------
+        images : Union[pa.StringArray, pa.BinaryArray, pa.FixedShapeTensorArray,
+            Iterable]
+
+        Returns
+        -------
+        Union[ImageURIArray, EncodedImageArray, FixedShapeImageTensorArray]
+        """
+
+        if isinstance(images, pa.StringArray):
+            return pa.ExtensionArray.from_storage(ImageURIType(), images)
+        elif isinstance(images, pa.BinaryArray):
+            return pa.ExtensionArray.from_storage(EncodedImageType(), images)
+        elif isinstance(images, pa.FixedShapeTensorArray):
+            shape = images.type.shape
+            value_type = images.type.value_type
+            typ = FixedShapeImageTensorType(value_type, shape)
+            return pa.ExtensionArray.from_storage(typ, images.storage)
+        elif isinstance(
+            images, (ImageURIArray, EncodedImageArray, FixedShapeImageTensorArray)
+        ):
+            return images
+        elif isinstance(images, (list, tuple, Iterable)):
+            return pa.ExtensionArray.from_storage(
+                ImageURIType(), pa.array(images, type=pa.string())
+            )
+
+        else:
+            raise TypeError("Cannot build a ImageArray from {}".format(type(images)))
+
+
+class ImageURIArray(ImageArray):
+    """
+    Array of image URIs. URIs may represent local files or remote files in
+    S3 or GCS if they are accessible by the machine executing this.
```

Comment:
Something for a follow up: it would be really cool if `https://` worked to pull public images off the web. Right now `read_uris()` gives the error:

```
ArrowInvalid: Unrecognized filesystem type in URI: https://lancedb.github.io/lance/_images/lance_logo.png
```

Reply:
I've added a very rudimentary download functionality to `read_uris` and removed uri checking from `from_uris`.
Exact change here: https://github.com/lancedb/lance/pull/1272/commits/8de58f900e24b345963b9c5aaaad633264e01fa2

Reply:
Is this decoding going to always run client-side? Otherwise we could be running arbitrary binaries through PIL and tensorflow decoders in private environments ..